[FIX] Rewrite the MRI/fieldmaps subsection for consistency with the rest of specs (#651)
bids-standard/bids-specification: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md

## Fieldmap data

Data acquired to correct for B0 inhomogeneities can come in different forms.

> The current version of this standard considers four different scenarios.

Reviewer: I suggest to briefly list the four scenario names already after this sentence.

The table that follows contains 5 rows and is a bit of a clash unless we are already introduced to the 4 scenarios.

## Fieldmap data

Data acquired to correct for *B<sub>0</sub>* inhomogeneities can come in different forms.
The current version of this standard considers four different scenarios:

 1. [Phase-difference map](#case-1-phase-difference-map-and-at-least-one-magnitude-image)
 1. [Two phase maps](#case-2-two-phase-maps-and-two-magnitude-images)
 1. [Direct *field mapping*](#case-3-direct-field-mapping)
 1. ["*PEpolar*" fieldmaps](#case-4-multiple-phase-encoded-directions-pepolar)

These four different types of field mapping strategies can be encoded
using the following image types:

| **Name**         | `suffix`         | **Description**                                                                                                                                                                                                      |
| ---------------- | ---------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | Magnitude        | magnitude\[1,2\] | Field-mapping MR schemes such as gradient-recalled echo (GRE) generate a Magnitude image to be used for anatomical reference. Requires the existence of Phase, Phase-difference or Fieldmap maps.                    |

Author: @effigies, what should we do about the brackets? just spell out the three options?

Reviewer 2: Oh, I think I misunderstood what was going on here. I think as long as it's clear from the text how they're used, then having two different notations seems fine as a way of flagging "check the text for details".

| Phase            | phase{1,2}       | Phase map generated by GRE schemes and alike, each associated with the first (`phase1`) or second (`phase2`) echoes in the sequence.                                                                                 |
| Phase-difference | phasediff        | Some scanners subtract the `phase1` from the `phase2` map and generate a unique `phasediff` file. For instance, this is a common output for the built-in fieldmap sequence of Siemens scanners.                      |
| Fieldmap         | fieldmap         | Some MR schemes such as spiral-echo (SE) sequences are able to directly provide maps of the B0 field inhomogeneity.                                                                                                  |
| EPI              | epi              | The phase-encoding polarity (PEpolar) technique combines two or more Spin Echo EPI scans with different phase encoding directions to estimate the underlying inhomogeneity/deformation map.                          |

Two OPTIONAL key/value pairs following more general rules of the specification
are allowed across all the four scenarios:

>   - The OPTIONAL [`run-<index>`](../99-appendices/09-entities.md#run) key/value pair correspond to a one-base index
>     to distinguish multiple fieldmaps with the same parameters.

Reviewer 3: ```suggestion
  - The OPTIONAL [`run-<index>`](../99-appendices/09-entities.md#run) key/value pair corresponds to a one-based index
    to distinguish multiple fieldmaps with the same parameters.
```

        sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_phasediff.nii[.gz]
        sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_phasediff.json
        sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_magnitude1.nii[.gz]

>         [sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_magnitude2.nii[.gz]]

Reviewer 3: Including brackets around the whole filename to indicate that it's an optional file isn't used consistently across the specification. I think it really only shows up in the MEG/EEG sections, so I'm not sure if we want to introduce it here.

Also, the schema isn't capable of storing that kind of information at the moment, so those brackets will be removed when/if #610 is merged.

Author: Not sure what the alternative would be. Any suggestions?

Reviewer: > Not sure what the alternative would be

I think it'd be fine not to communicate requirement level in the "template" section. After all, we are also not distinguishing between recommended and required files either. Users can see which files are optional from the spec text.

Author: I'm not particularly attached to this indication of an optional data object, but I think there are two different discussions here.

1. Whether the schema can store this information at the moment -- here the question is whether the schema should store this information, actually. The validator is going to know whether `magnitude2` must be found (phase1/2 fieldmaps) or is optional (phasediff fieldmaps).
2. Whether the template is the right place to state the indication

IMHO, 1) is harder to address, and outside the scope of this PR -- I guess that merging #610 will have to address this question, and if that means that the brackets in the template are gone because some other mechanism implements this not, I'm fine by that (that's how BIDS evolves, anyways).

Then 2) - I think that the Template is basically what most of the people check when preparing their data. Therefore, I think this is the most effective way of expressing that this element is optional. It also needs to be clearly stated within the text, but considering that magnitude2 may and may not be optional (phasediff and phase1/2, respectively), there is a larger chance this is lost in some future refactor. For clarity, I think having the template annotated with requirement levels, especially when irregular, is very useful.

@tsalo: would something like:

```
        sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_magnitude2.nii[.gz]  # OPTIONAL
```
work out?

Reviewer 3: There really isn't any mechanism for stating that some suffixes or extensions are optional in combination with others in the schema at the moment, so #610 will erase any indications of that from templates and users will need to rely on the text for that kind of information. I don't think it's that big a deal to address this issue in #610 though.

Author: What would you prefer then for this PR?

```
        [sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_magnitude2.nii[.gz]]
```
or
```
        sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_magnitude2.nii[.gz]  # OPTIONAL
```
I'm leaning towards the `# OPTIONAL` annotation which is more visible and obvious - WDYT?

Reviewer 3: I think `# OPTIONAL` works well.

        sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_phasediff.nii[.gz]
        sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_phasediff.json
        sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_magnitude1.nii[.gz]

>         sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_magnitude2.nii[.gz]  # OPTIONAL

Reviewer 3: One thing that confuses me about this template (even before this PR)- wouldn't we still have jsons for the magnitude maps?

Author: The magnitude maps are part of the acquisition (as the phase maps are), so all the metadata is probably best stored only on the phase information files. I'm not sure what metadata could possibly exist that only applies to the magnitude (and doesn't to the phase).

Reviewer 3: That's true, but I don't think that kind of relationship is supported in the specification. I guess we would call it something like horizontal inheritance?

I could be wrong though.

Reviewer 3: For tools like `pybids`, I think you need sidecar files for the magnitude images in order to extract their metadata. @effigies is that still the case or can `pybids` infer metadata from linked files?

Author: I haven't encountered any case where PyBIDS could not correctly query the metadata of fieldmaps.

        sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_phasediff.nii[.gz]
        sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_phasediff.json
        sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_magnitude1.nii[.gz]

>         sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_magnitude2.nii[.gz]  # OPTIONAL

Author: I think it is important to mark optional elements on the template. Following that argument all fieldmaps are essentially optional because a BIDS-valid dataset may or may not have them.

There's been some discussion about this annotation with @tsalo. Could you read through that and then let us know if you still think this change should be made (i.e., remove the OPTIONAL annotation in the template)?

>         sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_magnitude2.nii[.gz]  # OPTIONAL

Author: Also, I'm growing a feeling that `_sbref` should not exist altogether.

>         sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_magnitude2.nii[.gz]  # OPTIONAL

Reviewer 2: Agreed that all files are optional, so it seems strange to mark one out as super-extra optional in either form. (I did find and mark unresolved that comment thread, btw.)

Immediately below the template is: 

> where the REQUIRED _phasediff image corresponds to the phase-drift map between echo times, the REQUIRED _magnitude1 image corresponds to the shorter echo time, and the OPTIONAL _magnitude2 image to the longer echo time.

I don't understand why we need to add novel markup to the template when the text is right there.

Just as a final point, I would say that I'm not opposed to the concept of indicating in the template that one file in a group is optional, but I am opposed to introducing it in a one-off manner. I would rather we made a principled decision that takes into account all such groupings and applies it to the entire spec in one PR.

That said, I won't block this PR over this quibble.

>         sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_magnitude2.nii[.gz]  # OPTIONAL

Author: > I don't understand why we need to add novel markup to the template when the text is right there.

Most people will not read that paragraph, especially if they feel they know the specification and are making a quick confirmatory look-up.

> Just as a final point, I would say that I'm not opposed to the concept of indicating in the template that one file in a group is optional, but I am opposed to introducing it in a one-off manner. I would rather we made a principled decision that takes into account all such groupings and applies it to the entire spec in one PR.

I agree with this. For this PR I see three alternatives:

1. Leave it as is now and come back to the issue later (or within #610)
2. Roll it back to `[]` since that nomenclature, even though not formalized, is apparently present elsewhere in BIDS (and finally address it within #610).
3. Remove any annotations (and bring them up in a separate PR or #610).

I think 3 would be the best option, but I'm afraid that this will concept will be probably lost (I'm honestly not willing to go through the whole spec to propose something in a separate PR or in #610). So, after the discussion with @tsalo I think I prefer option 1.

       sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_fieldmap.json
```

Required fields:

| **Key name**   | **Requirement level** | **Data type**            | **Description**                                                                    |
| -------------- | --------------------- | ------------------------ | ---------------------------------------------------------------------------------- |
| Units          | REQUIRED              | [string][]               | Units of the fieldmap: Hertz (`Hz`), Radians per second (`rad/s`), or Tesla (`T`). |

For example:

```JSON
{
   "Units": "rad/s"
}
```

#### Case 4: Multiple phase encoded directions ("pepolar")
The phase-encoding polarity (PEpolar) technique combines two or more Spin Echo
EPI scans with different phase encoding directions to estimate the underlying
inhomogeneity/deformation map.

> These `_epi.nii[.gz]` files can be a 3D or 4D --

Reviewer 3: ```suggestion
These `_epi.nii[.gz]` files can be 3D or 4D --
```

       sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_fieldmap.json
```

In some cases (for example GE) the scanner software will output a precomputed
fieldmap denoting the B0 inhomogeneities along with a magnitude image used for
coregistration.
In this case the sidecar JSON file needs to include the units of the fieldmap.
The possible options are: Hertz (`Hz`), Radians per second (`rad/s`), or Tesla
(`T`).
Required fields:

| **Key name**   | **Requirement level** | **Data type**            | **Description**                                                                    |
| -------------- | --------------------- | ------------------------ | ---------------------------------------------------------------------------------- |
| Units          | REQUIRED              | [string][]               | Units of the fieldmap: Hertz (`Hz`), Radians per second (`rad/s`), or Tesla (`T`). |

For example:

```JSON
{
   "Units": "rad/s",

>    "IntendedFor": "func/sub-01_task-motor_bold.nii.gz"

Author: Yup, sorry about this.

>    "Units": "rad/s"

Author: ```suggestion
   "Units": "rad/s",
   "IntendedFor": "func/sub-01_task-motor_bold.nii.gz"
```

> 

Author: ```suggestion

See [Using `IntendedFor` metadata](#using-intendedfor-metadata)
for details on the `IntendedFor` field.

```

acquired for by filling the `IntendedFor` field in the corresponding JSON file.

| **Key name** | **Requirement level** | **Data type**                         | **Description**                                                                                                                                                                                                                                                                |
| ------------ | --------------------- | ------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| IntendedFor  | RECOMMENDED           | [string][] or [array][] of [string][] | Contains one or more filenames with paths relative to the participant subfolder. The path needs to use forward slashes instead of backward slashes. This field is OPTIONAL, and in case the fieldmaps do not correspond to any particular scans it does not have to be filled. |

For example:

```JSON
{

>    "IntendedFor": ["ses-pre/func/sub-01_ses-pre_task-motor_run-1_bold.nii.gz",
>                    "ses-post/func/sub-01_ses-post_task-motor_run-1_bold.nii.gz"]

Reviewer 3: This example seems unrealistic. Is there any real-world situation where people would acquire one field map in a longitudinal study and then apply it to fMRI scans from multiple sessions? I would lean toward a more common case.

```suggestion
   "IntendedFor": ["ses-pre/func/sub-01_ses-pre_task-motor_run-1_bold.nii.gz",
                   "ses-pre/func/sub-01_ses-pre_task-motor_run-2_bold.nii.gz"]
```

Author: Agreed. Sorry I saw your suggestion too late (and didn't commit your suggestion), but I came up with exactly the same example.

        sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_epi.json
```

> The [`dir-<label>`](../99-appendices/09-entities.md#dir) key/value pair is

Reviewer 3: ```suggestion
The [`dir-<label>`](../99-appendices/09-entities.md#dir) key/value pair used to
```

distinguish between different phase-encoding directions is REQUIRED for these
files, and SHOULD NOT be used to replace the required `PhaseEncodingDirection`
(nor any other acquisition parameter, as prescribed by the

> [common principles](../02-common-principles.md#file-name-structure).)

Reviewer 3: ```suggestion
[common principles](../02-common-principles.md#file-name-structure)).
```

        sub-<label>[_ses-<label>][_acq-<label>][_ce-<label>]_dir-<label>[_run-<index>]_epi.json
```

The [`dir-<label>`](../99-appendices/09-entities.md#dir) entity is REQUIRED
for these files.
This key-value pair SHOULD NOT be used to replace the required `PhaseEncodingDirection`

> (nor any other acquisition parameter, as prescribed by the
> [common principles](../02-common-principles.md#file-name-structure)).

Reviewer 2: ```suggestion
(see [File name structure](../02-common-principles.md#file-name-structure)).
```

Required fields:

| **Key name**           | **Requirement level** | **Data type** | **Description**                                                                    |
| ---------------------- | --------------------- | ------------- | ---------------------------------------------------------------------------------- |
| PhaseEncodingDirection | REQUIRED              | [string][]    | See [in-plane spatial encoding](#in-plane-spatial-encoding) table of fields.       |
| TotalReadoutTime       | REQUIRED              | [number][]    | See [in-plane spatial encoding](#in-plane-spatial-encoding) table of fields.       |

For example:

```JSON
{
   "PhaseEncodingDirection": "j-",

>    "TotalReadoutTime": 0.095

Author: ```suggestion
   "TotalReadoutTime": 0.095,
   "IntendedFor": "func/sub-01_task-motor_bold.nii.gz"
```

}
```

> 

Author: ```suggestion

See [Using `IntendedFor` metadata](#using-intendedfor-metadata)
for details on the `IntendedFor` field.

```

As for other EPI sequences, these field mapping sequences may have any of the
[in-plane spatial encoding](#in-plane-spatial-encoding) metadata keys.
However, please note that `PhaseEncodingDirection` and `TotalReadoutTime` keys
are REQUIRED for these field mapping sequences.

### Expressing the MR protocol intent for fieldmaps

Fieldmaps are typically acquired with the purpose of correcting one or more EPI
scans under `func/` or `dwi/` for distortions derived from *B<sub>0</sub>*
nonuniformity.
This linking between fieldmaps and their targetted data MAY be encoded with the
`IntendedFor` metadata.

> #### Using `IntendedFor` metadata

Reviewer 2: I guess the reason for dropping `IntendedFor` from earlier examples is because it's now introduced afterwards? I think this is generally confusing, as it breaks up fields that are going to appear together.

I'm not sure whether we should push this back up to the introductory section or just add:

```Markdown
See [Using `IntendedFor` metadata](#using-intendedfor-metadata)
for details on the `IntendedFor` field.
```

After each example.

Author: > I guess the reason for dropping `IntendedFor` from earlier examples is because it's now introduced afterwards?

Correct. I've edited as suggested.
